File: CN/JAVA/SlidingWindow_2.java
package JAVA;
import java.util.Scanner;

public class SlidingWindow_2 {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        // Prompt user for window size
        System.out.print("Enter the window size: ");
        int windowSize = scanner.nextInt();

        // Prompt user for number of frames to transmit
        System.out.print("Enter the number of frames to transmit: ");
        int numberOfFrames = scanner.nextInt();

        // Create an array to store the frames
        int[] frames = new int[numberOfFrames];
        
        // Allow user to enter the frames
        System.out.println("Enter the frames:");
        for (int i = 0; i < numberOfFrames; i++) {
            System.out.print("Frame " + (i + 1) + ": ");
            frames[i] = scanner.nextInt();
        }

        // Simulate sending frames in windows
        int ack = 0;
        while (ack < numberOfFrames) {
            // Send frames in the current window
            System.out.println("Sending frames:");
            for (int i = ack; i < ack + windowSize && i < numberOfFrames; i++) {
                System.out.println("Sent frame: " + frames[i]);
            }

            // Simulate waiting for acknowledgment
            System.out.println("Waiting for acknowledgment...");

            // In a real scenario, here we'd check for actual acknowledgments.
            // For simulation, assume all frames in the window are acknowledged.
            int acksReceived = Math.min(windowSize, numberOfFrames - ack);
            for (int i = 0; i < acksReceived; i++) {
                System.out.println("Acknowledgment received for frame: " + frames[ack + i]);
            }

            // Move the acknowledgment pointer
            ack += acksReceived;
        }

        System.out.println("All frames have been sent and acknowledged.");
        scanner.close();
    }
}

File: CN/Python/slidingWindow_2.py
# 2) Possible Question

"""Write a program to simulate the Sliding Window Protocol in a network transmission. Implement the following features:
1. Prompt the user to enter the window size.
2. Prompt the user to enter the number of frames to transmit.
3. Allow the user to enter the frames.
4. Simulate the sending of frames in windows, where the sender sends a window of frames and waits for an acknowledgment before sending the next window of frames.
5. Print the frames being sent and when the acknowledgment for those frames is received.
"""

def main():
    w = int(input("Enter window size: "))
    
    f = int(input("Enter number of frames to transmit: "))
    
    frames = []
    print(f"Enter {f} frames: ")
    for i in range(f):
        frames.append(int(input()))
    
    print("\nWith sliding window protocol, the frames will be sent in the following way (assuming no corruption of frames):")
    
    i = 0
    while i < f:
        print(f"\nSending frames: {frames[i:i+w]}")
        if (i + w) <= f:
            print("Acknowledgement of above frames sent is received by sender\n")
        else:
            print("Acknowledgement of above frames sent is received by sender\n")
        
        i += w

if __name__ == "__main__":
    main()